src/cascadia/TerminalSettingsModel/ActionArgs.idl

```diff
@@ -226,4 +233,17 @@ namespace Microsoft.Terminal.Settings.Model
         NewWindowArgs(NewTerminalArgs terminalArgs);
         NewTerminalArgs TerminalArgs { get; };
     };
+
+    unsealed runtimeclass SwitchToAdjacentTabArgs : IActionArgs
+    {
+        Windows.Foundation.IReference<TabSwitcherMode> SwitcherMode;
+    };
+
+    [default_interface] runtimeclass PrevTabArgs : SwitchToAdjacentTabArgs
+    {
+    };
+
+    [default_interface] runtimeclass NextTabArgs : SwitchToAdjacentTabArgs
+    {
+    };

```

Comment:
I would rather us not use WinRT composition for this. It's very ... um, not good.

It is implemented by the unsealed base class having an implementation for _every method_ that checks if it is being composed (inherited from).

It pretty much looks like this:

```
struct SwitchToAdjacentTabArgs {
    ISwitchToAdjacentTabArgsChild* _compositionChild;
    void Blah() {
        if (_compositionChild) { return _compositionChild->Blah(); }
        _Blah();
    }
};
```

with all of the weirdness required for reference tracking the composed child.

It also makes creating the composed child a pain, too. At runtime, that is. It has to allocate a whole SwitchToAdjacentTabArgs to be passed around in its place (or, in any place that ISwitchToAdjacentTabArgs can be used.)

Is there some way to structure this that is less unfortunate? :smile:

Reply:
There are several options (like using a non winrt type for sharing common logic), but I guess I will simply eliminate the inheritance.